appinventor/appengine/src/com/google/appinventor/server/BuildOutputServlet.java

```diff
@@ -95,6 +95,37 @@ public void doGet(HttpServletRequest req, HttpServletResponse resp) throws IOExc
         resp.sendError(resp.SC_NOT_FOUND, "Link has timed out");
         return;
       }
+
+      final String userId = nonce.getUserId();
+      final long projectId = nonce.getProjectId();
+
+      if (RemoteStorageInstanceHolder.isRemoteConfigured(RemoteStorageInstanceHolder.Usage.BUILD)) {
+        final StorageIo storageIo = StorageIoInstanceHolder.getInstance();
+        // Given all the output files, try to find the one that matches an output extension
+        final List<String> files = storageIo.getProjectOutputFiles(userId, projectId);
+        String outputFileName = null;
+        for (String fileName : files) {
+          if (BuildOutputFiles.isOutputFile(fileName)) {
+            outputFileName = new File(fileName).getName();
+            break;
+          }
+        }
+
+        if (outputFileName == null) {
+          throw new FileNotFoundException("No target file found!");
+        }
+
+        final RemoteStorage remoteStorage = RemoteStorageInstanceHolder.getInstance(RemoteStorageInstanceHolder.Usage.BUILD);
+        // For now, we use Android always. If we eventually support iOS, we should decide here
+        //   whether we are going for Android or Apple, as well as in the BarcodeAction.
+        final String target = BuildOutputFiles.getTargetName();
+        final String objectKey = remoteStorage.getBuildOutputObjectKey(target, userId, projectId, outputFileName);
+        final String remoteUrl = remoteStorage.generateRetrieveUrl(objectKey);
```

Comment:
Another option here would be to modify the Nonce, and along with the user and project, store a pre-calculated presigned URL, valid for as long as the Nonce is valid. This would remove generating in runtime the URL, and "make the URL consistent" if the same Nonce is accessed twice.

The only reason why I went with this approach to generate the URL on the fly is to not modify the current datastore schema, and as this is a local operation, it should be fine to spend a few ms more here. Accessing the same nonce twice is an unlikely situation too imo.